bdk_chain/src/sparse_chain.rs

```diff
@@ -317,60 +298,66 @@ impl<P: ChainPosition> SparseChain<P> {
-    pub fn insert_tx(&mut self, txid: Txid, pos: P) -> Result<bool, InsertTxErr> {
-        let new_height = pos.height();
-
-        let latest = self
-            .checkpoints
-            .keys()
-            .last()
-            .cloned()
-            .map(TxHeight::Confirmed);
-
-        if new_height.is_confirmed() && (latest.is_none() || new_height > latest.unwrap()) {
-            return Err(InsertTxErr::TxTooHigh);
+    /// Determines the resultant [`ChangeSet`] if [`Txid`] was inserted at position `pos`.
+    ///
+    /// Changes to the [`Txid`]'s position is allowed and will be reflected in the resultant
+    /// [`ChangeSet`].
+    pub fn insert_tx_changeset(
+        &self,
+        txid: Txid,
+        pos: P,
+    ) -> Result<ChangeSet<P>, UpdateFailure<P>> {
+        let mut update = Self::default();
+
+        if let Some(block_id) = self.latest_checkpoint() {
+            let _old_hash = update.checkpoints.insert(block_id.height, block_id.hash);
+            debug_assert!(_old_hash.is_none());
         }
 
-        if let Some(original_pos) = self.txid_to_pos.get(&txid) {
-            if original_pos.height().is_confirmed() && *original_pos != pos {
-                return Err(InsertTxErr::TxMoved);
+        let tip_height = self.checkpoints.iter().last().map(|(h, _)| *h);
+        if let TxHeight::Confirmed(tx_height) = pos.height() {
+            if Some(tx_height) > tip_height {
+                return Err(UpdateFailure::TxTooHigh {
+                    txid,
+                    tx_height,
+                    tip_height,
+                });
             }
-
-            return Ok(false);
         }
 
-        self.txid_to_pos.insert(txid, pos.clone());
-        self.ordered_txids.insert((pos, txid));
+        let _old_pos = update.txid_to_pos.insert(txid, pos.clone());
+        debug_assert!(_old_pos.is_none());
+
+        let _inserted = update.ordered_txids.insert((pos, txid));
+        debug_assert!(_inserted, "must insert tx");
 
-        Ok(true)
+        self.determine_changeset(&update)
     }
 
     /// Insert a `block_id` (a height and block hash) into the chain. The caller is responsible for
     /// guaranteeing that a block exists at that height. If a checkpoint already exists at that
-    /// height with a different hash this will return an error. Otherwise it will return `Ok(true)`
-    /// if the checkpoint didn't already exist or `Ok(false)` if it did.
-    ///
-    /// **Warning**: This function modifies the internal state of the chain. You are responsible
-    /// for persisting these changes to disk if you need to restore them.
-    pub fn insert_checkpoint(&mut self, block_id: BlockId) -> Result<bool, InsertCheckpointErr> {
-        if let Some(&old_hash) = self.checkpoints.get(&block_id.height) {
+    /// height with a different hash this will return an error.
+    pub fn insert_checkpoint_changeset(
+        &self,
+        block_id: BlockId,
+    ) -> Result<ChangeSet<P>, UpdateFailure<P>> {
```

Comment:
The error here should not change. There is only one thing that can go wrong which is `HashNotMatching`.